snapcraft/internal/project_loader/_config.py

```diff
@@ -71,6 +72,106 @@ def _validate_epoch(instance):
+            saw_strings = True
+        elif isinstance(mapping, dict):
+            saw_dicts = True
+            build_on = _get_architectures_list(mapping, 'build-on')
+            build_ons.update(build_on)
+
+            # Add to the list of run-ons. However, if no run-on is specified,
+            # we know it's implicitly the value of build-on, so use that
+            # for validation instead.
+            run_on = _get_architectures_list(mapping, 'run-on')
+            if run_on:
+                run_ons.update(run_on)
+            else:
+                standalone_build_ons.update(build_on)
+
+    # Architectures can either be a list of strings, or a list of objects.
+    # Mixing the two forms is unsupported.
+    if saw_strings and saw_dicts:
+        raise jsonschema.exceptions.ValidationError(
+            'every item must either be a string or an object',
+            path=['architectures'], instance=mapping)
+
+    # At this point, individual build-ons and run-ons have been validated,
+    # we just need to validate them across each other.
+
+    # First of all, if we have a `run-on: [all]` (or a standalone
+    # `build-on: [all]`) then we should only have one item in the instance,
+    # otherwise we know we'll have multiple snaps claiming they run on the same
+    # architectures (i.e. all and something else).
+    number_of_snaps = len(instance)
+    if 'all' in run_ons and number_of_snaps > 1:
+        raise jsonschema.exceptions.ValidationError(
+            "one of the items has 'all' in 'run-on', which only makes "
+            "sense if it's the only item, and there are {}".format(
+                number_of_snaps),
+            path=['architectures'], instance=instance)
+    if 'all' in build_ons and number_of_snaps > 1:
+        raise jsonschema.exceptions.ValidationError(
+            "one of the items has 'all' in 'build-on', which only makes "
+            "sense if it's the only item, and there are {}".format(
+                number_of_snaps),
+            path=['architectures'], instance=instance)
+
+    # We want to ensure that multiple `run-on`s (or standalone `build-on`s)
+    # don't incude the same arch, or they'll clash with each other when
+    # releasing.
+    all_ons = run_ons + standalone_build_ons
+    duplicates = {arch for (arch, count) in all_ons.items() if count > 1}
+    if duplicates:
+        raise jsonschema.exceptions.ValidationError(
+            'multiple items will build snaps that claim they run on {}'.format(
+                formatting_utils.humanize_list(duplicates, 'and')),
+            path=['architectures'], instance=instance)
+
+    # Finally, ensure that multiple `build-on`s don't include the same arch
+    # or Snapcraft has no way of knowing which one to use.
+    duplicates = {arch for (arch, count) in build_ons.items() if count > 1}
+    if duplicates:
+        raise jsonschema.exceptions.ValidationError(
+            'ambiguous run-ons when building on {}'.format(
```

Comment:
we need to expand this explanation, it is a bit confusing.

Reply:
Alright, I rephrased to be more explicit.